pkg/abi/abi.go

```diff
@@ -670,6 +670,63 @@ func (e *Entry) SignatureCtx(ctx context.Context) (string, error) {
 	return buff.String(), nil
 }
 
+func (e *Entry) SolidityDef() (string, []string, error) {
+	return e.SolidityDefCtx(context.Background())
+}
+
+// DescriptorStringCtx returns a Solidity-like descriptor of the entry, including its type
+func (e *Entry) SolidityDefCtx(ctx context.Context) (string, []string, error) {
+	// Everything apart from event and error is a type of function
+	isFunction := e.Type != Error && e.Type != Event
+
+	allChildStructs := []string{}
+	buff := new(strings.Builder)
+	buff.WriteString(string(e.Type))
+	buff.WriteRune(' ')
+	buff.WriteString(e.Name)
+	buff.WriteRune('(')
+	for i, p := range e.Inputs {
+		if i > 0 {
+			buff.WriteString(", ")
+		}
+		s, childStructs, err := p.SolidityDefCtx(ctx, isFunction)
+		if err != nil {
+			return "", nil, err
+		}
+		allChildStructs = append(allChildStructs, childStructs...)
+		buff.WriteString(s)
+	}
+	buff.WriteRune(')')
+
+	if isFunction {
+		buff.WriteString(" external")
+		if e.StateMutability != "" &&
+			// The state mutability nonpayable is reflected in Solidity by not specifying a state mutability modifier at all.
+			e.StateMutability != NonPayable {
+			buff.WriteRune(' ')
+			buff.WriteString(string(e.StateMutability))
+		}
+		if len(e.Outputs) > 0 {
+			buff.WriteString(" returns (")
+			for i, p := range e.Outputs {
+				if i > 0 {
+					buff.WriteString(", ")
+				}
+				s, childStructs, err := p.SolidityDefCtx(ctx, isFunction)
+				if err != nil {
+					return "", nil, err
+				}
+				allChildStructs = append(allChildStructs, childStructs...)
+				buff.WriteString(s)
+			}
+			buff.WriteRune(')')
+		}
+		buff.WriteString(" { }")
+	}
+
+	return buff.String(), allChildStructs, nil
```

Comment:
So we unpack and describe any structs at the top-level, but any nested structs are returned as a separate list?